### PR TITLE
BUG: MultiIndex.append not checking names for equality

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v1.6.0.rst [hunk 1]
```diff
@@ -161,7 +161,7 @@ Missing
 
 MultiIndex
 ^^^^^^^^^^
--
+- Bug in :meth:´MultiIndex.append` not checking names for equality (:issue:`48288`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.6.0.rst:164]:
```suggestion
- Bug in :meth:`MultiIndex.append` not checking names for equality (:issue:`48288`)
```

File: pandas/tests/indexes/multi/test_reshape.py [hunk 2]
```diff
@@ -150,6 +150,16 @@ def test_append_index():
     tm.assert_index_equal(result, expected)
 
 
+@pytest.mark.parametrize("name, exp", [("b", "b"), ("c", None)])
+def test_append_names_match(name, exp):
+    # GH#48288
+    midx = MultiIndex.from_arrays([[1, 2], [3, 4]], names=["a", "b"])
+    midx2 = MultiIndex.from_arrays([[3], [5]], names=["a", name])
```
Comment (mroeschke) [inline pandas/tests/indexes/multi/test_reshape.py:157]: So if names completely mismatch here e.g. `names=[name, "a"]` then the result names should be `[None, None]`? And do we have a test for this?

File: doc/source/whatsnew/v1.6.0.rst (hunk 1 above)
Comment (phofl) [reply doc/source/whatsnew/v1.6.0.rst]: thx

File: pandas/tests/indexes/multi/test_reshape.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/indexes/multi/test_reshape.py:157]: Yes, and no looks like I forgot this

**Comment by phofl** [comment]
Added test and also fixed a bug in pivot_table creating missmatches in Index names

File: pandas/core/indexes/multi.py [hunk 3]
```diff
@@ -2203,19 +2203,23 @@ def append(self, other):
...
             (isinstance(o, MultiIndex) and o.nlevels >= self.nlevels) for o in other
         ):
-            arrays = []
+            arrays, names = [], []
             for i in range(self.nlevels):
                 label = self._get_level_values(i)
                 appended = [o._get_level_values(i) for o in other]
                 arrays.append(label.append(appended))
-            return MultiIndex.from_arrays(arrays, names=self.names)
+                level_names = {label.name}.union({x.name for x in appended})
```
Comment (mroeschke) [inline pandas/core/indexes/multi.py:2211]: Could simplify to:

```
single_label_name = all(label.name == x.name for x in appended)
names.append(label.name if single_label_name else None)
```

so the name checking could short circuit if possible.

Comment (phofl) [reply pandas/core/indexes/multi.py:2211]: makes sense, thx

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**